bugfix/39 - Compass rose misaligned

wrksx opened the pull request:
Resolves #39
Resolves #1316

- Show and hide the compass marks around the view when selecting and deselecting an aircraft
- Align the compass rose to the selected aircraft

wrksx commented:
I just found the issue #1316, I fix it in this PR too, I didn't find this issue on time.
Should this PR also resolve #1316?
Let me know how to proceed, I can split this PR in two if necessary.

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1956,61 +1960,78 @@ export default class CanvasController {
+
+            // compute the intersection point between the ray and the canvas
+            const intersection = positive_intersection_with_rect(
+                aircraftPosition, rayUnitVector, canvasOrigin, canvasSize
```
Each parameter should be on its own line

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1956,61 +1960,78 @@ export default class CanvasController {
+            // major marks on headings multiple of 10 degrees
+            if (intersection) {
+                // set mark length and weight for standard, minor and major marks
+                let markLen;
```
Let’s default this to 8 and remove the else below

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1956,61 +1960,78 @@ export default class CanvasController {
+                } else {
+                    markLen = 8;
+                }
+                const markWeight = (heading % 30 === 0 ? 2 : 1);
```
* Missing empty line after `}`
* you can remove the pares here, they are nod necessary

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1956,61 +1960,78 @@ export default class CanvasController {
+            // standard marks on all headings
+            // minor marks on headings multiple of 5 degrees
+            // major marks on headings multiple of 10 degrees
+            if (intersection) {
```
Reverse the logic here to return early.  That way all the work of this method is not wrapped inside of an if statement

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -2032,18 +2053,42 @@ export default class CanvasController {
+        const vViewToScope = [CanvasStageModel._panX, CanvasStageModel._panY];
+        // the position of the view relative to the canvas origin
+        const vCanvasToView = [CanvasStageModel.halfWidth, CanvasStageModel.halfHeight];
+
```
Remove empty line

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -2032,18 +2053,42 @@ export default class CanvasController {
+            -CanvasStageModel.translateKilometersToPixels(positionFromScope[1])
         ];
+
+        // We want to compute a position relative to the canvas from a position relative to
```
This  is a great comment!  A better place for this is in the method doc block

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -2032,18 +2053,42 @@ export default class CanvasController {
-            CanvasStageModel.halfWidth + CanvasStageModel._panX + km(pos[0]),
-            CanvasStageModel.halfHeight + CanvasStageModel._panY - km(pos[1])
+    _toCanvasPosition(positionFromScope) {
+        // positionFromScope is given in kilometers so it needs to be translated
```
This is a great comment! A better place for this is in the method doc block

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -2250,6 +2295,13 @@ export default class CanvasController {
+    _onSelectAircraftStrip() {
+        this._markDeepRender();
+    }
+    _onDeselectAircraftStrip() {
```
Missing empty line between method declarations

n8rzz commented on src/assets/scripts/client/math/vector.js:
```diff
@@ -81,16 +81,39 @@ export const vnorm = (v, length) => {
+
+/**
+ * Alias of the vectorize_2d_from_radians
+ * Kept for backwards compatibility
```
😌

wrksx commented:
I fixed the code formatting according to your change request.
I also added tests for the new vector functions, which allowed me to catch an small approximation error that was causing a test to fail. Nice!

Regarding the tests, I'm not used to writing tests, specially for mathematical functions, I hope it'll do.
For the tests I needed a `is_within_epsilon()` function I that I added to the `vector.spec.js` file, but I feel it may not belong here...

n8rzz commented on src/assets/scripts/client/math/vector.js:
```diff
@@ -81,16 +81,46 @@ export const vnorm = (v, length) => {
+    while (heading_angle < 0) {
+        heading_angle += 2 * Math.PI;
+    }
+    while (heading_angle >= 2 * Math.PI) {
```
missing empty line above `while`

n8rzz commented on src/assets/scripts/client/math/vector.js:
```diff
@@ -81,16 +81,46 @@ export const vnorm = (v, length) => {
+export const vectorize_2d_from_radians = (heading_angle) => {
+    // keep the angle in the [0, 2*PI[ range
+    while (heading_angle < 0) {
+        heading_angle += 2 * Math.PI;
```
I'd encourage you to set `2 * Math.PI` to a variable so you only need to do the math once, and can then reference it every where else

n8rzz commented on src/assets/scripts/client/math/vector.js:
```diff
@@ -81,16 +81,46 @@ export const vnorm = (v, length) => {
+    while (heading_angle >= 2 * Math.PI) {
+        heading_angle -= 2 * Math.PI;
+    }
     return [
```
missing empty line above `return`

n8rzz commented on test/math/vector.spec.js:
```diff
@@ -2,13 +2,169 @@ import ava from 'ava';
+ *
+ * @return boolean  true if the value is within EPSILON of the expected value, false otherwise
+ */
+const is_within_epsilon = (value, expectedValue) => {
```
`is_within_epsilon` -> `isWithinEpsilon`

n8rzz commented on test/math/vector.spec.js:
```diff
@@ -2,13 +2,169 @@ import ava from 'ava';
+        return true;
+    }
+
+    return false;
```
this entire function could simply return your conditional:
```js
return value <= expectedValue + Number.EPSILON && value >= expectedValue - Number.EPSILON;
```

wrksx commented:
- Camel-cased and moved the `is_within_epsilon()` in `math/core.js`, right below a related and already existing `isWithin()` function. Moved the test to `core.spec.js`. That should dissolves any concern we had about testing a test utility function =).
- I wanted to reuse the `isWithin()` function but it was broken, had to fix it too. So I wrote tests for it while I was around.
- Camel-cased `vectorize_2d_from_radians()` and `vectorize_2d_from_degrees()`

I don't mind the nitpicking, that's how the codebase will get better!
Waiting for the review!

n8rzz commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -1956,64 +1960,86 @@ export default class CanvasController {
+        const canvasSize = [CanvasStageModel.width, CanvasStageModel.height];
+        const aircraftPosition = this._toCanvasPosition(aircraft.relativePosition);
+
+
```
remove extra empty space

n8rzz commented on src/assets/scripts/client/math/core.js:
```diff
@@ -84,12 +84,29 @@ export function s(i) {
-        const oldLimit1 = limit1;
-        limit1 = limit2;
-        limit2 = oldLimit1;
+        [limit2, limit1] = [limit1, limit2];
```
hmm, I don't remember seeing this one earlier.  what exactly are you doing here?

Are you creating an array (right side) but immediately destructuring to different assignments?  Either way, this is really confusing.  Is there a more obvious way to do what's needed here?

Clever solutions are fun, for sure, but they make things harder to maintain.  Simple is best 😉

wrksx replied:
That's exactly what you said. The goal is to swap variables, and that's the cleaner notation I've ever seen. I'm not usually fond of clever tricks either, but this is as explicit as we can get to swap vars.

We can restore it to what it was before, it was doing the job using a temporary variable `oldlimit1`:
```suggestion
        const oldLimit1 = limit1;
        limit1 = limit2;
        limit2 = oldLimit1;
```

n8rzz replied:
😌 

This is certainly more verbose, and perhaps not as fun, but this is much easier to read and understand quickly.  Thanks for changing this!

n8rzz commented on src/assets/scripts/client/math/core.js:
```diff
@@ -84,12 +84,29 @@ export function s(i) {
+ *
+ * @param {number}   value            the value we want to test
+ * @param {number}   expectedValue    the value we want to test against
+ *
```
remove extra line

n8rzz commented on test/math/core.spec.js:
```diff
@@ -1,11 +1,76 @@
+
+    // we will test providing the 'limit1' and 'limit2' both
+    // ways since the function is supposed to work that way
+    for (const number of tests) {
```
😌 so much more readable, thank you!

wrksx replied:
Not sure but I think your are mixing tests for `isWithin()` and `isWithinEpsilon()`. A probably contributing factor is that I used EPSILON in the test of `isWithin()` too.

Anyway since you like it better I rewrote the test for `isWithinEpsilon()` in the same style. And I agree a hundred percent it's way clearer now.

j-651 commented on src/assets/scripts/client/canvas/CanvasController.js:
```diff
@@ -2032,18 +2057,42 @@ export default class CanvasController {
+        // the position of the scope relative to the view (center of the canvas)
+        const vViewToScope = [CanvasStageModel._panX, CanvasStageModel._panY];
+        // the position of the view relative to the canvas origin (top left corner)
+        const vCanvasToView = [CanvasStageModel.halfWidth, CanvasStageModel.halfHeight];
```
Not sure if these are ideal, but I've renamed them to `scopePositionRelativeToView`, `viewPositionRelativeToCanvasOrigin`, and `scopePositionRelativeToCanvasOrigin`, respectively.

```js
const scopePositionRelativeToView = [CanvasStageModel._panX, CanvasStageModel._panY];
const viewPositionRelativeToCanvasOrigin = [CanvasStageModel.halfWidth, CanvasStageModel.halfHeight];
const scopePositionRelativeToCanvasOrigin = vadd(viewPositionRelativeToCanvasOrigin, scopePositionRelativeToView);
```
